File: Tepora-app/backend/src/core/em_llm/integrator.py
"""
EM-LLM integrator - Main integration point.

This module provides the EMLLMIntegrator class that orchestrates
all EM-LLM components for memory formation and retrieval.
"""

from __future__ import annotations

import asyncio
import logging
from typing import TYPE_CHECKING, Any, cast

import numpy as np

from .boundary import EMBoundaryRefiner
from .retrieval import EMTwoStageRetrieval
from .segmenter import EMEventSegmenter
from .types import EMConfig, EpisodicEvent

if TYPE_CHECKING:
    from ..embedding_provider import EmbeddingProvider
    from ..memory.memory_system import MemorySystem

logger = logging.getLogger(__name__)


class EMLLMIntegrator:
    """
    Integration class for EM-LLM with existing system.

    This class orchestrates:
    1. Event segmentation (surprise-based or semantic)
    2. Boundary refinement
    3. Representative token selection
    4. Memory storage and retrieval
    """

    def __init__(
        self,
        llm_diagnostics_provider: object | None,
        embedding_provider: EmbeddingProvider,
        config: EMConfig,
        memory_system: MemorySystem,
    ):
        """
        Initialize EM-LLM integrator.

        Args:
            llm_diagnostics_provider: Optional diagnostics provider for LLM config reporting
            embedding_provider: Provider for text embeddings
            config: EM-LLM configuration parameters
            memory_system: ChromaDB-backed memory system
        """
        self._llm_diagnostics_provider = llm_diagnostics_provider
        self.embedding_provider = embedding_provider
        self.config = config
        self.memory_system = memory_system

        # Initialize EM-LLM memory system
        self.segmenter = EMEventSegmenter(self.config)
        self.boundary_refiner = EMBoundaryRefiner(self.config)
        self.retrieval_system = EMTwoStageRetrieval(self.config, self.memory_system)

        logger.info("EM-LLM Integrator initialized")

    def get_current_llm_config_for_diagnostics(self) -> dict[str, Any]:
        """
        Get current LLM configuration for diagnostics.

        Returns:
            Dictionary with LLM configuration information
        """
        provider = self._llm_diagnostics_provider
        if provider and hasattr(provider, "get_current_model_config_for_diagnostics"):
            try:
                return cast(dict[str, Any], provider.get_current_model_config_for_diagnostics())
            except Exception:
                logger.debug("LLM diagnostics provider failed.", exc_info=True)
        return {}

    async def _finalize_and_store_events(self, events: list[EpisodicEvent]) -> list[EpisodicEvent]:
        """
        Common post-processing: select representative tokens, compute embeddings, store.

        Args:
            events: List of segmented events

        Returns:
            List of finalized events
        """
        if not events:
            return []

        # Select representative tokens and compute embeddings for each event
        for event in events:
            event.representative_tokens = self._select_representative_tokens(event)
            if self.embedding_provider and event.representative_tokens:
                repr_texts = [event.tokens[i] for i in event.representative_tokens]
                if repr_texts:
                    # Embedding computation is not async, so use asyncio.to_thread
                    embeddings = await asyncio.to_thread(self.embedding_provider.encode, repr_texts)
                    event.representative_embeddings = np.array(embeddings)

        # Store in memory
        self.retrieval_system.add_events(events)

        return events

    async def process_logprobs_for_memory(
        self, logprobs_content: list[dict]
    ) -> list[EpisodicEvent]:
        """
        Process LLM logprobs directly for surprise-based memory formation.

        This is the main memory formation path from the EM-LLM paper.

        Args:
            logprobs_content: List of logprob dictionaries from LLM

        Returns:
            List of created episodic events
        """
        logger.info("Processing logprobs for surprisal-based memory formation.")
        try:
            if not logprobs_content:
                logger.warning("logprobs_content is empty. Skipping memory formation.")
                return []

            # Normalize logprob entries
            normalized_entries = []
            for idx, entry in enumerate(logprobs_content):
                token = entry.get("token") or entry.get("token_str")
                logprob = entry.get("logprob")
                if token is None or logprob is None:
                    logger.debug(
                        "Skipping logprob entry %d due to missing token/logprob fields: %s",
                        idx,
                        entry,
                    )
                    continue
                normalized_entries.append({"token": token, "logprob": logprob})

            if not normalized_entries:
                logger.warning(
                    "No valid logprob entries after normalization. Skipping memory formation."
                )
                return []

            surprise_scores = self.segmenter.calculate_surprise_from_logprobs(normalized_entries)
            tokens = [item["token"] for item in normalized_entries]

            boundaries = self.segmenter._identify_event_boundaries(surprise_scores)

            events = []
            for i in range(len(boundaries) - 1):
                start, end = boundaries[i], boundaries[i + 1]
                event = EpisodicEvent(
                    tokens=tokens[start:end],
                    start_position=start,
                    end_position=end,
                    surprise_scores=surprise_scores[start:end],
                )
                events.append(event)

            # Boundary refinement (using attention keys if available)
            attention_keys = None  # Currently dummy
            if self.config.use_boundary_refinement and attention_keys is not None:
                logger.info("Applying boundary refinement using attention keys.")
                events = self.boundary_refiner.refine_boundaries(
                    events, attention_keys=attention_keys
                )

            return await self._finalize_and_store_events(events)
        except Exception as e:
            logger.error("EM-LLM logprobs processing failed: %s", e, exc_info=True)
            return []

    async def process_conversation_turn_for_memory(
        self, user_input: str, ai_response: str
    ) -> list[EpisodicEvent]:
        """
        Process conversation turn through EM-LLM memory formation pipeline.

        Args:
            user_input: User's input text
            ai_response: AI's response text

        Returns:
            List of created episodic events
        """
        logger.info(
            "Processing conversation turn for EM-LLM memory formation (semantic change based)."
        )

        try:
            if not ai_response:
                logger.warning("AI response is empty. Aborting memory formation.")
                return []

            # Semantic change-based memory formation pipeline
            logger.info("Processing text of %d chars for memory formation", len(ai_response))

            # Step 1 & 2: Semantic change-based segmentation
            events, sentence_embeddings = self.segmenter.segment_text_into_events(
                ai_response, self.embedding_provider
            )
            if not events:
                return []

            # Step 3: Boundary refinement
            if self.config.use_boundary_refinement and sentence_embeddings is not None:
                events = self.boundary_refiner.refine_boundaries(
                    events, context_vectors=sentence_embeddings, attention_keys=None
                )

            # Step 4 & 5: Common post-processing and memory storage
            final_events = await self._finalize_and_store_events(events)

            logger.info(
                "Created %d episodic events from conversation turn via semantic segmentation.",
                len(final_events),
            )
            return final_events

        except Exception as e:
            logger.error("EM-LLM memory formation failed: %s", e, exc_info=True)
            return []

    def _select_representative_tokens(self, event: EpisodicEvent) -> list[int]:
        """
        Select representative tokens within event (highest surprise scores).

        Args:
            event: Episodic event

        Returns:
            List of token indices (sorted)
        """
        if not event.surprise_scores:
            return []

        indexed_scores = [(score, i) for i, score in enumerate(event.surprise_scores)]
        indexed_scores.sort(key=lambda x: x[0], reverse=True)
        representative_indices = [i for _, i in indexed_scores[: self.config.repr_topk]]
        return sorted(representative_indices)

    def retrieve_relevant_memories_for_query(self, query: str) -> list[dict]:
        """
        Retrieve relevant memories using EM-LLM two-stage retrieval.

        Args:
            query: Query string

        Returns:
            List of memory entries (compatible with existing system)
        """
        logger.info("Retrieving memories using EM-LLM two-stage retrieval")

        try:
            if not self.embedding_provider:
                logger.warning("No embedding provider available for retrieval")
                return []

            # Execute EM-LLM two-stage retrieval
            query_embedding = np.array(self.embedding_provider.encode([query])[0])
            relevant_events = self.retrieval_system.retrieve_relevant_events(query_embedding)

            # Convert to dictionary format for compatibility with existing system
            memory_entries = []
            for i, event in enumerate(relevant_events):
                surprise_scores = event.surprise_scores or []
                surprise_mean = float(np.mean(surprise_scores)) if surprise_scores else 0.0
                surprise_max = float(np.max(surprise_scores)) if surprise_scores else 0.0
                event_size = len(event.tokens)

                memory_entry = {
                    "id": f"em_event_{event.start_position}_{event.end_position}",
                    "content": " ".join(event.tokens),
                    "summary": event.summary
                    or f"Episodic event from position {event.start_position} to {event.end_position}",
                    "surprise_stats": {
                        "mean_surprise": surprise_mean,
                        "max_surprise": surprise_max,
                        "event_size": event_size,
                    },
                    "representative_tokens": event.representative_tokens or [],
                    "retrieval_rank": i + 1,
                }
                memory_entries.append(memory_entry)

            logger.info("Retrieved %d EM-LLM memories", len(memory_entries))
            return memory_entries

        except AttributeError as e:
            logger.error(
                "EM-LLM memory retrieval failed due to missing component (e.g., embedding_provider): %s",
                e,
                exc_info=True,
            )
            return []
        except Exception as e:
            logger.error(
                "An unexpected error occurred during EM-LLM memory retrieval: %s", e, exc_info=True
            )
            return []

    def get_memory_statistics(self) -> dict:
        """
        Get current statistics of EM-LLM memory system.

        Returns:
            Dictionary with memory statistics
        """
        try:
            total_events = self.memory_system.count()

            stats = {
                "total_events": total_events,
                "configuration": {
                    "surprise_gamma": self.config.surprise_gamma,
                    "min_event_size": self.config.min_event_size,
                    "max_event_size": self.config.max_event_size,
                    "total_retrieved_events": self.config.total_retrieved_events,
                },
                "llm_config": self.get_current_llm_config_for_diagnostics(),
            }
            return stats
        except Exception as e:
            logger.error("Failed to get memory statistics: %s", e, exc_info=True)
            return {"status": "Error retrieving statistics", "error": str(e)}

File: Tepora-app/backend-rs/src/em_llm/integrator.rs
//! EM-LLM integrator - Main integration point.
//!
//! This module provides the EMLLMIntegrator class that orchestrates
//! all EM-LLM components for memory formation and retrieval.

use crate::em_llm::{
    EMBoundaryRefiner, EMConfig, EMEventSegmenter, EMTwoStageRetrieval, EpisodicEvent,
};

/// Main integration class for the EM-LLM system.
///
/// Orchestrates:
/// 1. Event segmentation (surprise-based or semantic)
/// 2. Boundary refinement (graph-theoretic)
/// 3. Memory storage and retrieval (two-stage)
pub struct EMLLMIntegrator {
    config: EMConfig,
    segmenter: EMEventSegmenter,
    refiner: EMBoundaryRefiner,
    retrieval: EMTwoStageRetrieval,
}

impl EMLLMIntegrator {
    /// Create a new EM-LLM integrator with the given configuration.
    pub fn new(config: EMConfig) -> Self {
        let segmenter = EMEventSegmenter::new(config.clone());
        let refiner = EMBoundaryRefiner::new(config.clone());
        let retrieval = EMTwoStageRetrieval::new(config.clone());

        Self {
            config,
            segmenter,
            refiner,
            retrieval,
        }
    }

    /// Create with default configuration.
    pub fn default() -> Self {
        Self::new(EMConfig::default())
    }

    /// Get current configuration.
    pub fn config(&self) -> &EMConfig {
        &self.config
    }

    /// Process LLM logprobs for surprise-based memory formation.
    ///
    /// This is the main memory formation path from the EM-LLM paper.
    ///
    /// # Arguments
    /// * `logprobs` - List of (token, logprob) pairs from LLM output
    /// * `embeddings` - Optional embedding vectors for boundary refinement
    ///
    /// # Returns
    /// List of created episodic events
    pub fn process_logprobs_for_memory(
        &mut self,
        logprobs: &[(String, f64)],
        embeddings: Option<&[Vec<f32>]>,
    ) -> Vec<EpisodicEvent> {
        if logprobs.is_empty() {
            return vec![];
        }

        // Extract tokens and calculate surprise scores
        let tokens: Vec<String> = logprobs.iter().map(|(t, _)| t.clone()).collect();
        let surprise_scores = self.segmenter.calculate_surprise_from_logprobs(logprobs);

        // Segment into events
        let mut events = self.segmenter.segment_tokens(&tokens, &surprise_scores);

        // Apply boundary refinement if embeddings provided
        if let Some(embs) = embeddings {
            if self.config.use_boundary_refinement {
                events = self.refiner.refine_boundaries(events, embs);
            }
        }

        // Select representative tokens and store events
        let finalized_events = self.finalize_events(events, embeddings);

        // Store in retrieval system
        self.retrieval.add_events(finalized_events.clone());

        finalized_events
    }

    /// Process a conversation turn for memory formation.
    ///
    /// Alternative path for when logprobs are not available.
    /// Uses semantic embeddings to detect event boundaries.
    ///
    /// # Arguments
    /// * `sentences` - List of sentences from the conversation
    /// * `embeddings` - Embedding vector for each sentence
    ///
    /// # Returns
    /// List of created episodic events
    pub fn process_conversation_for_memory(
        &mut self,
        sentences: &[String],
        embeddings: &[Vec<f32>],
    ) -> Vec<EpisodicEvent> {
        if sentences.is_empty() || sentences.len() != embeddings.len() {
            return vec![];
        }

        // Segment by semantic change
        let boundaries = self
            .segmenter
            .segment_by_semantic_change(sentences, embeddings);

        // Create events from boundaries
        let mut events = Vec::new();
        for window in boundaries.windows(2) {
            let start = window[0];
            let end = window[1];

            let event_sentences = sentences[start..end].to_vec();
            let summary = event_sentences.join(" ");

            // Create pseudo surprise scores (semantic distance-based)
            let surprise_scores = vec![1.0; event_sentences.len()];

            let mut event = EpisodicEvent::new(
                uuid::Uuid::new_v4().to_string(),
                event_sentences,
                start,
                end,
                surprise_scores,
            );
            event.summary = Some(summary);

            // Use mean embedding for the event
            if start < embeddings.len() && end <= embeddings.len() {
                let event_embeddings = &embeddings[start..end];
                event.embedding = Some(mean_embedding(event_embeddings));
            }

            events.push(event);
        }

        // Handle last segment
        if let Some(&last_boundary) = boundaries.last() {
            if last_boundary < sentences.len() {
                let event_sentences = sentences[last_boundary..].to_vec();
                let summary = event_sentences.join(" ");
                let surprise_scores = vec![1.0; event_sentences.len()];

                let mut event = EpisodicEvent::new(
                    uuid::Uuid::new_v4().to_string(),
                    event_sentences,
                    last_boundary,
                    sentences.len(),
                    surprise_scores,
                );
                event.summary = Some(summary);

                if last_boundary < embeddings.len() {
                    let event_embeddings = &embeddings[last_boundary..];
                    event.embedding = Some(mean_embedding(event_embeddings));
                }

                events.push(event);
            }
        }

        // Store in retrieval system
        self.retrieval.add_events(events.clone());

        events
    }

    /// Retrieve relevant memories for a query.
    ///
    /// Uses two-stage retrieval (similarity + contiguity).
    ///
    /// # Arguments
    /// * `query_embedding` - Embedding vector for the query
    ///
    /// # Returns
    /// List of relevant episodic events
    pub fn retrieve_memories(&self, query_embedding: &[f32]) -> Vec<EpisodicEvent> {
        self.retrieval.retrieve(query_embedding)
    }

    /// Retrieve memories with custom number of results.
    pub fn retrieve_memories_with_k(
        &self,
        query_embedding: &[f32],
        k: usize,
    ) -> Vec<EpisodicEvent> {
        self.retrieval.retrieve_with_k(query_embedding, Some(k))
    }

    /// Get memory statistics.
    pub fn get_statistics(&self) -> MemoryStatistics {
        MemoryStatistics {
            total_events: self.retrieval.event_count(),
            config: self.config.clone(),
        }
    }

    /// Finalize events by selecting representative tokens and computing embeddings.
    fn finalize_events(
        &self,
        mut events: Vec<EpisodicEvent>,
        embeddings: Option<&[Vec<f32>]>,
    ) -> Vec<EpisodicEvent> {
        for event in &mut events {
            // Select representative tokens (highest surprise)
            let representative = self.select_representative_tokens(event);
            event.representative_tokens = Some(representative);

            // Generate summary from tokens
            event.summary = Some(event.text());

            // Compute event embedding from token embeddings
            if let Some(embs) = embeddings {
                // Get embeddings for this event's token range
                let start = event.start_position;
                let end = event.end_position.min(embs.len());
                if start < end {
                    let event_embeddings = &embs[start..end];
                    event.embedding = Some(mean_embedding(event_embeddings));
                }
            }
        }

        events
    }

    /// Select representative tokens within an event (highest surprise scores).
    fn select_representative_tokens(&self, event: &EpisodicEvent) -> Vec<usize> {
        let topk = self.config.repr_topk.min(event.surprise_scores.len());

        // Get indices sorted by surprise score (descending)
        let mut indexed_scores: Vec<(usize, f64)> = event
            .surprise_scores
            .iter()
            .enumerate()
            .map(|(i, &s)| (i, s))
            .collect();

        indexed_scores.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

        // Return top-k indices, sorted by position
        let mut top_indices: Vec<usize> = indexed_scores
            .into_iter()
            .take(topk)
            .map(|(i, _)| i)
            .collect();
        top_indices.sort();
        top_indices
    }
}

/// Statistics about the EM-LLM memory system.
#[derive(Debug, Clone)]
pub struct MemoryStatistics {
    pub total_events: usize,
    pub config: EMConfig,
}

/// Calculate mean embedding from a list of embeddings.
fn mean_embedding(embeddings: &[Vec<f32>]) -> Vec<f32> {
    if embeddings.is_empty() {
        return vec![];
    }

    let dim = embeddings[0].len();
    let n = embeddings.len() as f32;

    let mut mean = vec![0.0f32; dim];
    for emb in embeddings {
        for (i, &val) in emb.iter().enumerate() {
            if i < dim {
                mean[i] += val / n;
            }
        }
    }

    mean
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_integrator_creation() {
        let integrator = EMLLMIntegrator::default();
        assert_eq!(integrator.config().surprise_window, 128);
    }

    #[test]
    fn test_logprobs_processing() {
        let mut integrator = EMLLMIntegrator::new(EMConfig {
            min_event_size: 2,
            max_event_size: 5,
            ..Default::default()
        });

        // Create test logprobs with a spike
        let logprobs: Vec<(String, f64)> = vec![
            ("Hello".to_string(), -0.5),
            (" ".to_string(), -0.1),
            ("world".to_string(), -0.3),
            ("!".to_string(), -0.2),
            (" ".to_string(), -0.1),
            ("How".to_string(), -3.0), // Spike
            (" ".to_string(), -0.1),
            ("are".to_string(), -0.4),
            (" ".to_string(), -0.1),
            ("you".to_string(), -0.3),
        ];

        let events = integrator.process_logprobs_for_memory(&logprobs, None);

        // Should have created at least one event
        assert!(!events.is_empty());

        // Statistics should reflect stored events
        let stats = integrator.get_statistics();
        assert!(stats.total_events > 0);
    }

    #[test]
    fn test_memory_retrieval() {
        let mut integrator = EMLLMIntegrator::default();

        // Add events with embeddings
        let sentences = vec![
            "The sky is blue.".to_string(),
            "The grass is green.".to_string(),
            "The sun is bright.".to_string(),
        ];
        let embeddings = vec![
            vec![1.0, 0.0, 0.0],
            vec![0.9, 0.1, 0.0],
            vec![0.0, 1.0, 0.0],
        ];

        integrator.process_conversation_for_memory(&sentences, &embeddings);

        // Query similar to first sentences
        let query = vec![1.0, 0.0, 0.0];
        let results = integrator.retrieve_memories(&query);

        // Should retrieve something
        assert!(!results.is_empty());
    }
}
